### PR TITLE
1.4 Update

### PR DESCRIPTION
**Pull request by AlexejheroYTB**

### CODE REVIEW

File: QModManager/Patcher.cs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+
+namespace QModManager
+{
+    public class QModPatcher
+    {
+        internal static string QModBaseDir = Environment.CurrentDirectory.Contains("system32") && Environment.CurrentDirectory.Contains("Windows") ? "ERR" : Path.Combine(Environment.CurrentDirectory, "QMods");
```
[the remark this answers is not in the document]
Comment (AlexejheroYTB) [reply QModManager/Patcher.cs]: Sometimes `Environment.CurrentDirectory` return `C:\Windows\system32`, that's why I used this check. Also no, I intended to use `&&`.

[the remark this answers is not in the document]
Comment (PrimeSonic) [reply QModManager/Patcher.cs]: 🆗

File: QModManager/Utility.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Reflection;
+
+namespace QModManager.Utility
+{
+    public static class ExceptionUtils
+    {
+        public static void OutputInnerExceptionRecursively(Exception e)
```
[the remark this answers is not in the document]
Comment (AlexejheroYTB) [reply QModManager/Utility.cs]: File removed

File: QModManager/Patcher.cs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        {
+            try
+            {
+                LoadMods();
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine("EXCEPTION CAUGHT!");
+                Console.WriteLine(e.Message);
+                Console.WriteLine(e.StackTrace);
```
[the remark this answers is not in the document]
Comment (AlexejheroYTB) [reply QModManager/Patcher.cs]: Addressed in commit 7b2c639

File: QModManager/Injector.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                MethodDefinition patchMethod = installer.MainModule.GetType("QModInstaller.QModPatcher").Methods.First(x => x.Name == "Patch");
+
+                TypeDefinition type = game.MainModule.GetType("GameInput");
+                MethodDefinition method = type.Methods.Single(x => x.Name == "Awake");
+
+                method.Body.GetILProcessor().InsertBefore(method.Body.Instructions[0], Instruction.Create(OpCodes.Call, method.Module.Import(patchMethod)));
+
+                game.Write(mainFilename);
+
+                if (!Directory.Exists(Path.Combine(gameDirectory, "QMods"))) Directory.CreateDirectory(Path.Combine(gameDirectory, "QMods"));
```
[the remark this answers is not in the document]
Comment (AlexejheroYTB) [reply QModManager/Injector.cs]: Addressed in commit 6cd4254

File: QModManager/QMod.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        public static QMod FromJsonFile(string file)
+        {
+            try
+            {
+                JsonSerializerSettings settings = new JsonSerializerSettings
+                {
+                    MissingMemberHandling = MissingMemberHandling.Ignore,
+                };
+
+                string json = File.ReadAllText(file);
```
[the remark this answers is not in the document]
Comment (PrimeSonic) [reply QModManager/QMod.cs]: 🆗

**Pull request merged**